lib/llhttp/parser.js

```diff
@@ -0,0 +1,217 @@
+  const statusMessage = i[kStatusMessage]
+  const upgrade = Boolean(inst.exports.llhttp_get_upgrade(p))
+  const shouldKeepAlive = inst.exports.llhttp_should_keep_alive(p)
+
+  return i[kOnHeadersComplete](versionMajor, versionMinor, rawHeaders, null,
+    null, statusCode, statusMessage, upgrade, shouldKeepAlive)
+}
+
+const wasm_on_body = (p, at, length) => {
+  const i = instMap.get(p)
+  const u8 = new Uint8Array(inst.exports.memory.buffer, at, length)
+  const body = Buffer.from(u8) // llhttp re-uses buffer so we need to make a copy.
+  i[kOnBody](body)
+  return 0
+}
+
+const wasm_on_message_complete = (p) => {
+  const i = instMap.get(p)
+  // Handle trailers
+  if (i[kHeadersFields].length) {
+    const rawHeaders = i[kGetHeaders]()
+    i[kOnHeaders](rawHeaders)
+  }
+  return instMap.get(p)[kOnMessageComplete]()
+}
+
+/* eslint-enable camelcase */
+const instMap = new Map()
+
+const inst = new WebAssembly.Instance(mod, {
+  env: {
+    wasm_on_message_begin,
+    wasm_on_url,
+    wasm_on_status,
+    wasm_on_header_field,
+    wasm_on_header_value,
+    wasm_on_headers_complete,
+    wasm_on_body,
+    wasm_on_message_complete
+  }
+})
+
+inst.exports._initialize() // wasi reactor
+
+class HTTPParserError extends Error {
+  constructor (message, code) {
+    super(message)
+    Error.captureStackTrace(this, HTTPParserError)
+    this.name = 'HTTPParserError'
+    this.code = code
+  }
+}
+
+class HTTPParser {
+  constructor (maxHeaderSize, lenient, headersTimeout) {
+    this[kPtr] = inst.exports.llhttp_alloc(constants.TYPE.RESPONSE)
+    this[kBufferSize] = 0
+    this[kBufferPtr] = null
+    this[kStatusMessage] = null
+    this[kHeadersFields] = []
+    this[kHeadersValues] = []
+    this[kMaxHeadersSize] = maxHeaderSize
+    this[kHeadersTimeout] = headersTimeout
+    instMap.set(this[kPtr], this)
+  }
+
+  [kMallocBuffer] (size) {
+    if (this[kBufferPtr]) {
+      inst.exports.free(this[kBufferPtr])
+    }
+    this[kBufferSize] = size
+    this[kBufferPtr] = inst.exports.malloc(size)
+  }
+
+  [kGetHeaders] () {
+    const rawHeaders = []
+    for (let c = 0; c < this[kHeadersFields].length; c++) {
+      rawHeaders.push(this[kHeadersFields][c], this[kHeadersValues][c])
+    }
+    this[kHeadersFields] = []
+    this[kHeadersValues] = []
+    return rawHeaders
+  }
+
+  [kReset] () {
+    this[kStatusMessage] = null
+  }
+
+  [kOnMessageBegin] () {
+    return 0
+  }
+
+  [kOnHeaders] (rawHeaders) {}
+
+  [kOnHeadersComplete] (versionMajor, versionMinor, rawHeaders, method,
+    url, statusCode, statusMessage, upgrade, shouldKeepAlive) {
+    return 0
+  }
+
+  [kOnBody] (body) {
+    return 0
+  }
+
+  [kOnMessageComplete] () {
+    return 0
+  }
+
+  close () {
+    instMap.delete(this[kPtr])
+    inst.exports.llhttp_free(this[kPtr])
+    inst.exports.llhttp_free(this[kBufferPtr])
+  }
+
+  execute (data) {
+    if (data.byteLength > this[kBufferSize]) {
+      this[kMallocBuffer](data.byteLength)
```

Comment:
```suggestion
      this[kMallocBuffer](Math.ceil(data.byteLength / 4096) * 4096)
```

Reply:
Just bc I don't know, why the `ceil` calculation?

Reply:
so that we don't cause a re allocation because of small difference in buffer size, e.g. you wouldn't want to reallocate in the following case:

```js
128
222
312
313
400
600
etc...
```

Reply:
Great, thank you.